// @theme-ui/mdx/dist/index.modern.js
import { jsx } from '@theme-ui/core';
import { css, get } from '@theme-ui/css';
import styled from '@emotion/styled';
import { useMDXComponents, MDXProvider as MDXProvider$1 } from '@mdx-js/react';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

const tags = ['p', 'b', 'i', 'a', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'img', 'pre', 'code', 'ol', 'ul', 'li', 'blockquote', 'hr', 'em', 'table', 'tr', 'th', 'td', 'em', 'strong', 'del', // mdx
'inlineCode', 'thematicBreak', // other
'div', // theme-ui
'root'];
const aliases = {
  inlineCode: 'code',
  thematicBreak: 'hr',
  root: 'div'
};

const isAlias = x => x in aliases;

const alias = n => isAlias(n) ? aliases[n] : n;

const themed = key => props => css(get(props.theme, `styles.${key}`))(props.theme);
const Styled = styled('div')(themed('div'));
const components = {};
tags.forEach(tag => {
  // fixme?
  components[tag] = styled(alias(tag))(themed(tag));
  Styled[tag] = components[tag];
});

const createComponents = comps => {
  const next = _extends({}, components);

  const componentKeys = Object.keys(comps);
  componentKeys.forEach(key => {
    next[key] = styled(comps[key])(themed(key));
  });
  return next;
};

const MDXProvider = ({
  components,
  children
}) => {
  const outer = useMDXComponents();
  return jsx(MDXProvider$1, {
    components: createComponents(_extends({}, outer, components)),
    children
  });
};

export { MDXProvider, Styled, components, themed };
